File: Microcam/Monitoring/ActivityMonitor.swift
import AppKit
@preconcurrency import ApplicationServices
import Combine
import CoreGraphics
import Foundation

@MainActor
final class ActivityMonitor: NSObject, ObservableObject, @unchecked Sendable {
    @Published private(set) var displayState: MonitorDisplayState = .stopped
    @Published private(set) var accessibilityGranted = AXIsProcessTrusted()
    private(set) var currentSegment: ActivitySegment?
    @Published private(set) var pauseUntil: Date?
    @Published private(set) var lastRecordedAt: Date?
    @Published private(set) var lastPersistenceError: String?

    var isPaused: Bool {
        if case .paused = displayState { return true }
        return false
    }

    var onDidPersist: (@MainActor () -> Void)?
    var onWake: (@MainActor () -> Void)?
    var onPeriodicMaintenance: (@MainActor () -> Void)?

    private let settings: SettingsStore
    private let store: SQLiteStore
    private var timer: Timer?
    private var accessibilityRefreshTask: Task<Void, Never>?
    private var lastAccessibilityCheck = Date.distantPast
    private var lastTitleFallbackCheck = Date.distantPast
    private var isStarted = false
    private var sessionSuspended = false
    private var manuallyPaused = false

    private var axObserver: AXObserver?
    private var axApplication: AXUIElement?
    private var axWindow: AXUIElement?
    private var observedPID: pid_t?

    init(settings: SettingsStore, store: SQLiteStore) {
        self.settings = settings
        self.store = store
        super.init()
    }

    func start() async {
        guard !isStarted else { return }
        isStarted = true
        installWorkspaceObservers()

        // Application switches and window-title changes are event driven. This timer is
        // only a low-frequency safety net for idle detection, midnight splitting and
        // crash-recovery checkpoints.
        let timer = Timer(timeInterval: 60, repeats: true) { [weak self] _ in
            Task { @MainActor [weak self] in
                await self?.tick()
            }
        }
        timer.tolerance = 15
        RunLoop.main.add(timer, forMode: .common)
        self.timer = timer

        accessibilityGranted = AXIsProcessTrusted()
        await evaluateCurrentActivity(forceTransition: true)
    }

    func stop() async {
        guard isStarted else { return }
        isStarted = false
        timer?.invalidate()
        timer = nil
        accessibilityRefreshTask?.cancel()
        accessibilityRefreshTask = nil
        removeWorkspaceObservers()
        removeAccessibilityObserver()
        await closeCurrent(at: Date())
        setDisplayState(.stopped)
    }

    func pause(for interval: TimeInterval?) async {
        manuallyPaused = interval == nil
        pauseUntil = interval.map { Date().addingTimeInterval($0) }
        await closeCurrent(at: Date())
        setDisplayState(.paused(until: pauseUntil))
    }

    func resume() async {
        manuallyPaused = false
        pauseUntil = nil
        await evaluateCurrentActivity(forceTransition: true)
    }

    func requestAccessibilityPermission() {
        let options = [
            kAXTrustedCheckOptionPrompt.takeUnretainedValue() as String: true
        ] as CFDictionary
        _ = AXIsProcessTrustedWithOptions(options)
        // The system prompt is asynchronous. Watch briefly only after an explicit user action
        // so the UI reacts immediately without adding a permanent high-frequency poll.
        watchForAccessibilityPermissionChange()
    }

    func watchForAccessibilityPermissionChange() {
        accessibilityRefreshTask?.cancel()
        accessibilityRefreshTask = Task { @MainActor [weak self] in
            for _ in 0..<60 {
                guard let self, !Task.isCancelled else { return }
                await refreshAccessibilityStatus()
                if accessibilityGranted { return }
                do {
                    try await Task.sleep(for: .seconds(1))
                } catch {
                    return
                }
            }
        }
    }

    func refreshAccessibilityStatus() async {
        lastAccessibilityCheck = Date()
        let newValue = AXIsProcessTrusted()
        guard newValue != accessibilityGranted else { return }
        accessibilityGranted = newValue
        await evaluateCurrentActivity(forceTransition: true)
    }

    fileprivate func accessibilityEventReceived(focusedWindowChanged: Bool) async {
        guard isStarted else { return }
        if focusedWindowChanged {
            configureAccessibilityObserver(for: NSWorkspace.shared.frontmostApplication, force: true)
        }
        await evaluateCurrentActivity(forceTransition: false)
    }

    private func tick() async {
        guard isStarted else { return }
        let now = Date()

        if let pauseUntil, now >= pauseUntil {
            self.pauseUntil = nil
        }
        if now.timeIntervalSince(lastAccessibilityCheck) >= 300 {
            lastAccessibilityCheck = now
            await refreshAccessibilityStatus()
        }
        let shouldRefreshTitle = now.timeIntervalSince(lastTitleFallbackCheck) >= 300
        if shouldRefreshTitle { lastTitleFallbackCheck = now }
        await evaluateCurrentActivity(
            forceTransition: false,
            refreshWindowTitle: shouldRefreshTitle
        )
        onPeriodicMaintenance?()
    }

    private func evaluateCurrentActivity(
        forceTransition: Bool,
        refreshWindowTitle: Bool = true
    ) async {
        let now = Date()

        guard settings.recordingEnabled else {
            await closeCurrent(at: now)
            setDisplayState(.stopped)
            return
        }
        guard !sessionSuspended else { return }
        if manuallyPaused || pauseUntil != nil {
            await closeCurrent(at: now)
            setDisplayState(.paused(until: pauseUntil))
            return
        }

        await splitAtMidnightIfNeeded(now: now)

        let anyInputEvent = CGEventType(rawValue: UInt32.max) ?? .null
        let idleSeconds = CGEventSource.secondsSinceLastEventType(
            .combinedSessionState,
            eventType: anyInputEvent
        )
        let threshold = TimeInterval(settings.idleMinutes * 60)
        if idleSeconds >= threshold {
            let idleBoundary = now.addingTimeInterval(-(idleSeconds - threshold))
            await closeCurrent(at: idleBoundary)
            setDisplayState(.idle)
            return
        }

        guard let application = NSWorkspace.shared.frontmostApplication else {
            await closeCurrent(at: now)
            setDisplayState(.idle)
            return
        }
        let bundleID = application.bundleIdentifier ?? "pid.\(application.processIdentifier)"
        let appName = application.localizedName ?? "未知应用"
        let requestedPolicy = settings.policy(for: bundleID)

        if requestedPolicy == .exclude {
            await closeCurrent(at: now)
            setDisplayState(.recording(appName: "已排除 \(appName)"))
            removeAccessibilityObserver()
            return
        }

        let effectivePolicy: AppCapturePolicy = requestedPolicy == .title && !accessibilityGranted
            ? .durationOnly
            : requestedPolicy

        if effectivePolicy == .title {
            configureAccessibilityObserver(for: application)
        } else {
            removeAccessibilityObserver()
        }

        // The raw title is scoped to this expression and is never persisted or logged.
        let canReuseCurrentTitle = !refreshWindowTitle
            && currentSegment?.bundleID == bundleID
            && currentSegment?.capturePolicy == effectivePolicy
        let sanitizedTitle: String?
        if effectivePolicy != .title {
            sanitizedTitle = nil
        } else if canReuseCurrentTitle {
            sanitizedTitle = currentSegment?.sanitizedTitle
        } else {
            lastTitleFallbackCheck = now
            sanitizedTitle = settings.redactor.redact(readFocusedWindowTitle(for: application))
        }

        let changed = forceTransition ||
            currentSegment?.bundleID != bundleID ||
            currentSegment?.sanitizedTitle != sanitizedTitle ||
            currentSegment?.capturePolicy != effectivePolicy

        if changed {
            await closeCurrent(at: now)
            let start = max(now.addingTimeInterval(-idleSeconds), Calendar.autoupdatingCurrent.startOfDay(for: now))
            currentSegment = ActivitySegment(
                id: UUID(),
                startAt: start,
                endAt: now,
                bundleID: bundleID,
                appName: appName,
                sanitizedTitle: sanitizedTitle,
                capturePolicy: effectivePolicy
            )
        } else {
            currentSegment?.endAt = now
        }

        if let currentSegment {
            do {
                try await store.save(currentSegment)
                markPersistenceSucceeded(at: now)
                onDidPersist?()
            } catch {
                // No private values are included in user-facing or system logs.
                setPersistenceError(error.localizedDescription)
                setDisplayState(.stopped)
                return
            }
        }

        setDisplayState(
            requestedPolicy == .title && !accessibilityGranted
                ? .permissionLimited
                : .recording(appName: appName)
        )
    }

    private func closeCurrent(at requestedEnd: Date) async {
        guard var segment = currentSegment else { return }
        let end = max(segment.startAt, min(requestedEnd, Date()))
        segment.endAt = end
        currentSegment = nil
        guard segment.activeSeconds > 0 else { return }
        do {
            try await store.save(segment)
            markPersistenceSucceeded(at: Date())
            onDidPersist?()
        } catch {
            setPersistenceError(error.localizedDescription)
            setDisplayState(.stopped)
        }
    }

    private func setDisplayState(_ newValue: MonitorDisplayState) {
        guard displayState != newValue else { return }
        displayState = newValue
    }

    private func markPersistenceSucceeded(at date: Date) {
        if lastPersistenceError != nil { lastPersistenceError = nil }
        if lastRecordedAt == nil || date.timeIntervalSince(lastRecordedAt ?? .distantPast) >= 60 {
            lastRecordedAt = date
        }
    }

    private func setPersistenceError(_ message: String) {
        guard lastPersistenceError != message else { return }
        lastPersistenceError = message
    }

    private func splitAtMidnightIfNeeded(now: Date) async {
        guard let segment = currentSegment else { return }
        let startOfToday = Calendar.autoupdatingCurrent.startOfDay(for: now)
        guard segment.startAt < startOfToday else { return }
        await closeCurrent(at: startOfToday)
    }

    private func readFocusedWindowTitle(for application: NSRunningApplication) -> String? {
        let appElement = AXUIElementCreateApplication(application.processIdentifier)
        var windowValue: CFTypeRef?
        guard AXUIElementCopyAttributeValue(
            appElement,
            kAXFocusedWindowAttribute as CFString,
            &windowValue
        ) == .success, let windowValue else { return nil }

        let window = unsafeDowncast(windowValue, to: AXUIElement.self)
        var titleValue: CFTypeRef?
        guard AXUIElementCopyAttributeValue(
            window,
            kAXTitleAttribute as CFString,
            &titleValue
        ) == .success else { return nil }
        return titleValue as? String
    }

    private func configureAccessibilityObserver(for application: NSRunningApplication?, force: Bool = false) {
        guard
            accessibilityGranted,
            let application,
            force || observedPID != application.processIdentifier
        else { return }

        removeAccessibilityObserver()
        let appElement = AXUIElementCreateApplication(application.processIdentifier)
        var observer: AXObserver?
        guard AXObserverCreate(
            application.processIdentifier,
            microcamAccessibilityCallback,
            &observer
        ) == .success, let observer else { return }

        let context = Unmanaged.passUnretained(self).toOpaque()
        _ = AXObserverAddNotification(
            observer,
            appElement,
            kAXFocusedWindowChangedNotification as CFString,
            context
        )

        var windowValue: CFTypeRef?
        if AXUIElementCopyAttributeValue(
            appElement,
            kAXFocusedWindowAttribute as CFString,
            &windowValue
        ) == .success, let windowValue {
            let window = unsafeDowncast(windowValue, to: AXUIElement.self)
            _ = AXObserverAddNotification(
                observer,
                window,
                kAXTitleChangedNotification as CFString,
                context
            )
            axWindow = window
        }

        CFRunLoopAddSource(
            CFRunLoopGetMain(),
            AXObserverGetRunLoopSource(observer),
            .commonModes
        )
        axApplication = appElement
        axObserver = observer
        observedPID = application.processIdentifier
    }

    private func removeAccessibilityObserver() {
        if let axObserver {
            CFRunLoopRemoveSource(
                CFRunLoopGetMain(),
                AXObserverGetRunLoopSource(axObserver),
                .commonModes
            )
        }
        axObserver = nil
        axApplication = nil
        axWindow = nil
        observedPID = nil
    }

    private func installWorkspaceObservers() {
        let center = NSWorkspace.shared.notificationCenter
        center.addObserver(self, selector: #selector(applicationActivated), name: NSWorkspace.didActivateApplicationNotification, object: nil)
        center.addObserver(self, selector: #selector(sessionResigned), name: NSWorkspace.sessionDidResignActiveNotification, object: nil)
        center.addObserver(self, selector: #selector(sessionBecameActive), name: NSWorkspace.sessionDidBecomeActiveNotification, object: nil)
        center.addObserver(self, selector: #selector(systemWillSleep), name: NSWorkspace.willSleepNotification, object: nil)
        center.addObserver(self, selector: #selector(systemDidWake), name: NSWorkspace.didWakeNotification, object: nil)
        center.addObserver(self, selector: #selector(screensDidSleep), name: NSWorkspace.screensDidSleepNotification, object: nil)
        center.addObserver(self, selector: #selector(screensDidWake), name: NSWorkspace.screensDidWakeNotification, object: nil)
        NotificationCenter.default.addObserver(
            self,
            selector: #selector(microcamBecameActive),
            name: NSApplication.didBecomeActiveNotification,
            object: nil
        )
    }

    private func removeWorkspaceObservers() {
        NSWorkspace.shared.notificationCenter.removeObserver(self)
        NotificationCenter.default.removeObserver(self, name: NSApplication.didBecomeActiveNotification, object: nil)
    }

    @objc private func applicationActivated(_: Notification) {
        Task { @MainActor [weak self] in
            await self?.evaluateCurrentActivity(forceTransition: true)
        }
    }

    @objc private func microcamBecameActive(_: Notification) {
        Task { @MainActor [weak self] in
            guard let self else { return }
            await refreshAccessibilityStatus()
            await evaluateCurrentActivity(forceTransition: false)
        }
    }

    @objc private func sessionResigned(_: Notification) {
        Task { @MainActor [weak self] in
            guard let self else { return }
            sessionSuspended = true
            await closeCurrent(at: Date())
        }
    }

    @objc private func sessionBecameActive(_: Notification) {
        Task { @MainActor [weak self] in
            guard let self else { return }
            sessionSuspended = false
            await evaluateCurrentActivity(forceTransition: true)
            onWake?()
        }
    }

    @objc private func systemWillSleep(_: Notification) { sessionResigned(Notification(name: NSWorkspace.willSleepNotification)) }
    @objc private func screensDidSleep(_: Notification) { sessionResigned(Notification(name: NSWorkspace.screensDidSleepNotification)) }
    @objc private func systemDidWake(_: Notification) { sessionBecameActive(Notification(name: NSWorkspace.didWakeNotification)) }
    @objc private func screensDidWake(_: Notification) { sessionBecameActive(Notification(name: NSWorkspace.screensDidWakeNotification)) }
}

private func microcamAccessibilityCallback(
    _: AXObserver,
    _: AXUIElement,
    notification: CFString,
    refcon: UnsafeMutableRawPointer?
) {
    guard let refcon else { return }
    let monitor = Unmanaged<ActivityMonitor>.fromOpaque(refcon).takeUnretainedValue()
    let focusedWindowChanged = CFEqual(
        notification,
        kAXFocusedWindowChangedNotification as CFString
    )
    Task { @MainActor in
        await monitor.accessibilityEventReceived(focusedWindowChanged: focusedWindowChanged)
    }
}

File: windows/src/main/activity-monitor.js
import { EventEmitter } from "node:events";
import os from "node:os";
import { Redactor } from "../core/redactor.js";
import { SENSITIVE_EXECUTABLES } from "../core/defaults.js";
import { dayBounds, localDayString } from "../core/time.js";

const CHECKPOINT_INTERVAL = 30_000;

export class ActivityMonitor extends EventEmitter {
  constructor({ settingsStore, database, collector }) {
    super();
    this.settingsStore = settingsStore;
    this.database = database;
    this.collector = collector;
    this.current = null;
    this.manualPause = false;
    this.pauseUntil = null;
    this.sessionSuspended = false;
    this.lastPersistedAt = null;
    this.lastCheckpoint = 0;
    this.state = process.platform === "win32" ? "stopped" : "platform-unsupported";
    this.error = null;
    this.redactor = null;
    this.refreshRedactor();
    collector.on("snapshot", (snapshot) => this.#handleSnapshot(snapshot));
    collector.on("error", (error) => this.#setError(error.message));
    collector.on("unsupported", () => this.#setState("platform-unsupported"));
  }

  start() {
    this.collector.start();
    if (process.platform === "win32") this.#setState("recording");
  }

  stop() {
    this.collector.stop();
    this.closeCurrent(Date.now());
    this.#setState("stopped");
  }

  suspend() {
    this.sessionSuspended = true;
    this.closeCurrent(Date.now());
    this.#setState("suspended");
  }

  resumeFromSuspend() {
    this.sessionSuspended = false;
    if (!this.manualPause && !this.pauseUntil) this.#setState("recording");
  }

  pause(minutes = null) {
    this.manualPause = minutes === null;
    this.pauseUntil = minutes === null ? null : Date.now() + Math.max(1, Number(minutes)) * 60_000;
    this.closeCurrent(Date.now());
    this.#setState("paused");
  }

  resume() {
    this.manualPause = false;
    this.pauseUntil = null;
    this.#setState(this.settingsStore.value.recordingEnabled ? "recording" : "stopped");
  }

  refreshRedactor() {
    const settings = this.settingsStore.snapshot();
    this.redactor = new Redactor({
      customTerms: settings.customSensitiveTerms,
      customPatterns: settings.customPatterns,
      username: os.userInfo().username
    });
  }

  snapshot() {
    return {
      state: this.state,
      currentApp: this.current?.appName ?? null,
      pauseUntil: this.pauseUntil,
      lastRecordedAt: this.lastPersistedAt,
      error: this.error
    };
  }

  closeCurrent(requestedEnd) {
    if (!this.current) return;
    const segment = this.current;
    this.current = null;
    segment.endAt = Math.max(segment.startAt, Math.min(requestedEnd, Date.now()));
    if (segment.endAt > segment.startAt) this.#persist(segment);
  }

  #handleSnapshot(snapshot) {
    const now = Number(snapshot.capturedAt) || Date.now();
    const settings = this.settingsStore.value;
    if (this.pauseUntil && now >= this.pauseUntil) this.resume();

    if (this.sessionSuspended) return;

    if (!settings.recordingEnabled) {
      this.closeCurrent(now);
      this.#setState("stopped");
      return;
    }
    if (this.manualPause || this.pauseUntil) {
      this.closeCurrent(now);
      this.#setState("paused");
      return;
    }
    if (snapshot.unavailable || !snapshot.executable) {
      this.closeCurrent(now);
      this.#setState("unavailable");
      return;
    }

    if (this.current && localDayString(this.current.startAt) !== localDayString(now)) {
      const boundary = dayBounds(localDayString(this.current.startAt)).end;
      this.closeCurrent(boundary);
    }

    const idleSeconds = Math.max(0, Number(snapshot.idleSeconds) || 0);
    const thresholdSeconds = settings.idleMinutes * 60;
    if (idleSeconds >= thresholdSeconds) {
      const idleBoundary = now - ((idleSeconds - thresholdSeconds) * 1000);
      this.closeCurrent(idleBoundary);
      this.#setState("idle");
      return;
    }

    const appId = String(snapshot.executable).toLowerCase();
    const appName = String(snapshot.appName || snapshot.executable).slice(0, 160);
    const configuredPolicy = this.settingsStore.policyFor(appId);
    const policy = configuredPolicy ?? (SENSITIVE_EXECUTABLES.has(appId) ? "durationOnly" : "title");
    if (policy === "exclude") {
      this.closeCurrent(now);
      this.#setState("excluded");
      return;
    }

    // The raw title is scoped to this expression and never logged or persisted.
    const sanitizedTitle = policy === "title" ? this.redactor.redact(snapshot.title) : null;
    const changed = !this.current || this.current.appId !== appId ||
      this.current.capturePolicy !== policy || this.current.sanitizedTitle !== sanitizedTitle;
    if (changed) {
      this.closeCurrent(now);
      this.current = this.database.createSegment({
        startAt: now,
        endAt: now,
        appId,
        appName,
        sanitizedTitle,
        capturePolicy: policy
      });
    } else {
      this.current.endAt = now;
    }

    if (now - this.lastCheckpoint >= CHECKPOINT_INTERVAL && this.current.endAt > this.current.startAt) {
      this.#persist(this.current);
    }
    this.error = null;
    this.#setState("recording");
  }

  #persist(segment) {
    try {
      this.database.saveSegment(segment);
      this.lastPersistedAt = Date.now();
      this.lastCheckpoint = this.lastPersistedAt;
      this.emit("persisted");
    } catch {
      this.#setError("活动日志写入失败");
    }
  }

  #setError(message) {
    this.error = message;
    this.#setState("error");
  }

  #setState(value) {
    if (this.state === value) return;
    this.state = value;
    this.emit("changed", this.snapshot());
  }
}
